[Nyayapati Srinivasa Nikhil] iP

nnmoq opened the pull request:
### Duke

> “Your mind is for having ideas, not holding them.” – _David Allen_ ([source](https://dansilvestre.com/productivity-quotes))

Duke is a simple task list. It's

- text-based
- easy to use
- ~~FAST~~ SUPER FAST to learn

All you need to do is,

1. Download it from [here](https://github.com/nnmoq/ip/releases/download/A-Jar/Duke.jar) and store it in a new directory
2. Start terminal and go to the directory Duke.jar is located.
3. Run command `java -jar Duke.jar`
4. Let it manage your tasks. 😉 

And it is **FREE**!

Features

- [x] Managing Tasks
- [x] Managing Deadlines
- [x] Managing Events

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
    public static void main(String[] args) {
        Duke chatBot = new Duke("data/data.txt");
        chatBot.run();
    }
```

Commit:
Add Gradle support

nnmoq committed:
Finished Level-1

nnmoq committed:
Finished Level-2

nnmoq committed:
Finished Level-3 and optional unmark

nnmoq committed:
Changed to switch statement.

nnmoq committed:
Completed Level-4.

nnmoq committed:
Completed A-TextUiTesting.

nnmoq committed:
Completed Level-5.

nnmoq committed:
Completed Level-6 and updated testcases.

nnmoq committed:
Add save functionality.

Save functionality is very useful as it can store the task list after Duke
closes.

nnmoq committed:
Add date functionality.

Date functionality is very useful as it can format date for easy readability.

nnmoq committed:
branch-Level-7 merged with master.

nnmoq committed:
Merge branch-Level-8

nnmoq committed:
Fix Deadline format bug.

Saved Deadline in data file causes date format issue.

nnmoq committed:
Use more OOP

Abstract more classes.

nnmoq committed:
Package classes.

Package classes together so they are organised.

nnmoq committed:
Add custom exceptions and bug fixes.

Custom exceptions and bug fixes are needed to prevent runtime crashes.

nnmoq committed:
Print only once using  ui.

nnmoq committed:
Update test cases.

nnmoq committed:
Add JUnit tests.

JUnit tests will automate testing in an isolated environment.

nnmoq committed:
Delete Main and move to Duke

nnmoq committed:
Add find functionality.

find [keyword] will allow search functionality.

nnmoq committed:
Add javaDocs.

Java documentation allows easy readability of the purpose of the methods/classes.

nnmoq committed:
Follow coding standard.

Tweak the coding standard to the given coding standard for easy readability.

nnmoq committed:
Merge JavaDocs to master branch.

nnmoq committed:
Merge Level-9 branch to master.

nnmoq committed:
Bug fix.

nnmoq committed:
Merge remote-tracking branch 'origin/add-gradle-support'

nnmoq committed:
Revert "Bug fix."

This reverts commit 1dba56ec8b4cbbd507678a3c9c1fc65fc0d58845.

nnmoq committed:
Update test cases.

nnmoq committed:
Bug fixes.

Make InvalidDateException show proper date format.
Minor coding style fixes.

nnmoq committed:
Add gradle checkStyle.

Conform to coding standards.

nnmoq committed:
Abstract all print statements to Ui.

nnmoq committed:
Style Duke GUI and fix GUI to exit when duke exits.

nnmoq committed:
Update javadocs.

Spyobird commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String printTask() {
+        if (this.isMarked) {
+            return "[" + type + "][X] " + activity + " (by "
+                    + date.format(DateTimeFormatter.ofPattern("MMM dd yyyy")) + ")";
+        } else {
+            return "[" + type + "][ ] " + activity + " (by "
+                    + date.format(DateTimeFormatter.ofPattern("MMM dd yyyy")) + ")";
+        }
+    }
```
Perhaps you could consider simplifying this with some abstraction?

nnmoq replied:
Thanks, I've brought out  `date.format(DateTimeFormatter.ofPattern("MMM dd yyyy"))` as a constant.

Spyobird reviewed (commented):
Overall very well written. Good use of OOP principles and easy to understand code. Very intuitive implementation of this program. Well done!

TeddYE reviewed (commented):
Overall, I found your code easy to read.

TeddYE commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,32 @@
+    /**
+     * Constructor for the exit command.
+     */
+    public ExitCommand(){
```
Shouldn't there be a white space before the open curly bracket?

nnmoq replied:
Yep, fixed it.

TeddYE commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public class Event extends Task {
+
+    private String at;
```
Shouldn't the variable name be more descriptive? Instead of at maybe you can try eventDateTime?

nnmoq replied:
I changed it to `eventLocation`.

TeddYE commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,171 @@
+public class Parser {
+
+    // Global constant commands.
+    private static final SimpleDateFormat taskFormat = new SimpleDateFormat("MMM dd yyyy");
```
Shouldn't constant names be all uppercase using underscore to separate words?

nnmoq replied:
Yep changed it to `TASK_FORMAT` and `SAVED_FORMAT`.

nnmoq committed:
Bug fix on Parser

There are several cases where "command body" where body is missing causes a runtime exception making the application crash.

Fixing these will make duke more robust.

As a step towards improving code quality, let's catch these exceptions. Doing so will make duke more robust.

nnmoq committed:
Add assertions

There are several cases where a null value will cause a runtime exception making the application crash.

As a step towards improving robustness, let's add assertions. Doing so will make duke stop running instantly instead of further down the line making debugging easier.

nnmoq committed:
Improve Code Quality.

The Parser class is hard to read and there are some minor bugs in the code.

As a step towards improving code quality, let's make more abstractions. Doing so will make the Parser class more reader friendly improving the code quality.

nnmoq committed:
Merge pull request #1 from nnmoq/branch-A-CodeQuality

Improve Code Quality.

nnmoq committed:
Merge branch 'master' into branch-A-Assertions

nnmoq committed:
Merge pull request #2 from nnmoq/branch-A-Assertions

Add assertions

nnmoq committed:
Add more assertions.

nnmoq committed:
Improve code quality.

nnmoq committed:
Add support for B-DoAfterTasks.

nnmoq committed:
Add CI

nnmoq committed:
Checkstyle Fix

nnmoq committed:
Checkstyle Fix for CI pass.

nnmoq committed:
Javadocs header comments confer to checkstyle.

nnmoq committed:
Make GUI more pleasant.

nnmoq committed:
Add credits.

nnmoq committed:
Add README.md and Ui.png

nnmoq committed:
Update README.md

nnmoq committed:
Remove redundant CLI tests.

nnmoq committed:
Update README.md with start up image.

nnmoq committed:
Fix date format in README.md

nnmoq committed:
Set theme jekyll-theme-merlot

nnmoq committed:
Update UI.png

nnmoq committed:
Merge remote-tracking branch 'origin/master'

nnmoq committed:
Revert dialogBox color for error.

nnmoq committed:
Update GUI style

nnmoq committed:
Update Ui and StartUp to new GUI style

nnmoq committed:
Pass checkstyle

nnmoq committed:
Conform variable name to checkstyle

nnmoq committed:
Update quick setup
